tonic/src/transport/channel/service/reconnect.rs

```diff
@@ -186,31 +218,42 @@ enum Inner<F> {
 }
 
 impl<F> ResponseFuture<F> {
-    pub(crate) fn new(inner: F) -> Self {
+    pub(crate) fn new(inner: F, reconnect: ReconnectNotify) -> Self {
         ResponseFuture {
             inner: Inner::Future(inner),
+            reconnect,
         }
     }
 
     pub(crate) fn error(error: crate::BoxError) -> Self {
         ResponseFuture {
             inner: Inner::Error(Some(error)),
+            reconnect: ReconnectNotify::new(),
         }
     }
 }
 
-impl<F, T, E> Future for ResponseFuture<F>
+impl<F, E> Future for ResponseFuture<F>
 where
-    F: Future<Output = Result<T, E>>,
+    F: Future<Output = Result<http::Response<Body>, E>>,
     E: Into<crate::BoxError>,
 {
-    type Output = Result<T, crate::BoxError>;
+    type Output = Result<http::Response<Body>, crate::BoxError>;
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        //self.project().inner.poll(cx).map_err(Into::into)
         let me = self.project();
         match me.inner.project() {
-            InnerProj::Future(fut) => fut.poll(cx).map_err(Into::into),
+            InnerProj::Future(fut) => match ready!(fut.poll(cx)) {
+                Ok(res) => {
+                    // Server errors map to `Status::UNAVAILABLE` and are unrecoverable
+                    if res.status().is_server_error() {
```

Comment:
This doesn't feel correct, what if you are connected to a proxy and want to keep the connection alive but the downstream from the proxy returned a 500 but it was transient. So now you did all that work to reconnect even though keeping the connection would have been better?

Reply:
My understanding is that if we receive a non-successful code then we should be expecting a 200 status with `grpc-header` set.

Other status codes would imply that we have lost the gRPC connection.

To put it into context, TCP proxies do not implement gRPC and would not be expected know to signal transient conditions using 500 error codes.